src/consensus/encode.rs

```diff
@@ -547,6 +547,38 @@ impl Encodable for [u16; 8] {
     }
 }
 
+/// Read wrapper capped to `remaining` bytes (default: [MAX_VEC_SIZE])
+pub struct CappedRead<'r> {
+    reader: &'r mut dyn io::Read,
+    remaining: usize,
+}
+
+impl<'r> CappedRead<'r> {
+    /// New [CappedRead] from a Read type capped to [MAX_VEC_SIZE] bytes
+    pub fn new<R: io::Read>(reader: &'r mut R) -> Self {
+        Self::with_cap(reader, MAX_VEC_SIZE)
+    }
+
+    /// New [CappedRead] from a Read type with custom max bytes
+    pub fn with_cap<R: io::Read>(reader: &'r mut R, cap: usize) -> Self {
+        CappedRead {
+            reader: reader as &'r mut dyn io::Read,
+            remaining: cap,
+        }
+    }
+}
+
+impl<'r> io::Read for CappedRead<'r> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let bytes = self.reader.read(buf)?;
```

Comment:
It seems to me that `buf` should be sliced here: before being passed to underlying `reader` so that the `reader` is never even asked for more than the limit. Otherwise the `reader` could e.g. hang waiting for more data than the limit, which is not desireable.

Reply:
true, but not applicable anymore because we now use `take()`